File: mezzanine-compiler/src/main/java/com/anthonycr/mezzanine/MezzanineProcessor.java
package com.anthonycr.mezzanine;

import com.anthonycr.mezzanine.filter.SupportedElementFilter;
import com.anthonycr.mezzanine.generator.MezzanineGenerator;
import com.anthonycr.mezzanine.map.MezzanineMapper;
import com.anthonycr.mezzanine.source.ElementSource;
import com.anthonycr.mezzanine.source.MezzanineElementSource;
import com.anthonycr.mezzanine.utils.MessagerUtils;
import com.google.auto.service.AutoService;

import java.util.Set;

import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.Processor;
import javax.annotation.processing.RoundEnvironment;
import javax.annotation.processing.SupportedAnnotationTypes;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.TypeElement;

@AutoService(Processor.class)
@SupportedAnnotationTypes(value = {"com.anthonycr.mezzanine.FileStream"})
public class MezzanineProcessor extends AbstractProcessor {

    private boolean isProcessed;

    @Override
    public SourceVersion getSupportedSourceVersion() {
        return SourceVersion.latestSupported();
    }

    @Override
    public boolean process(Set<? extends TypeElement> set, RoundEnvironment roundEnvironment) {
        if (isProcessed) {
            return true;
        }

        isProcessed = true;

        MessagerUtils.initialize(processingEnv.getMessager());

        ElementSource mezzanineElementSource = new MezzanineElementSource(roundEnvironment);

        MessagerUtils.reportInfo("Starting Mezzanine processing");

        mezzanineElementSource.createElementStream()
            .filter(SupportedElementFilter.filterForSupportedElements())
            .map(MezzanineMapper.elementToTypeAndFilePair())
            .doOnNext(typeElementFileEntry -> MessagerUtils.reportInfo("Processing file: " + typeElementFileEntry.getValue()))
            .map(MezzanineMapper.fileToStringContents())
            .map(MezzanineGenerator.generateGeneratorTypeSpecs())
            .toList()
            .map(MezzanineGenerator.generateMezzanineTypeSpec())
            .map(MezzanineGenerator.generateJavaFile())
            .flatMapCompletable(MezzanineGenerator.writeFileToDisk(processingEnv.getFiler()))
            .subscribe(() -> MessagerUtils.reportInfo("File successfully processed"));

        return true;
    }
}

File: mezzanine-compiler/src/main/java/com/anthonycr/mezzanine/MezzanineProcessor.kt
package com.anthonycr.mezzanine

import com.anthonycr.mezzanine.extensions.doOnNext
import com.anthonycr.mezzanine.filter.SupportedElementFilter
import com.anthonycr.mezzanine.function.ElementToTypeAndFilePairFunction
import com.anthonycr.mezzanine.function.GenerateFileStreamTypeSpecFunction
import com.anthonycr.mezzanine.function.GenerateMezzanineTypeSpecFunction
import com.anthonycr.mezzanine.function.TypeSpecToJavaFileFunction
import com.anthonycr.mezzanine.utils.MessagerUtils
import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.processing.SymbolProcessor
import com.google.devtools.ksp.symbol.KSAnnotated
import com.squareup.kotlinpoet.ksp.writeTo

class MezzanineProcessor(
    private val codeGenerator: CodeGenerator,
    private val logger: KSPLogger,
) : SymbolProcessor {
    override fun process(resolver: Resolver): List<KSAnnotated> {
        MessagerUtils.messager = logger

        MessagerUtils.reportInfo("Starting processing")

        val symbolsWithAnnotation =
            resolver.getSymbolsWithAnnotation(FileStream::class.qualifiedName!!)

        val fileStreamTypeSpecs = symbolsWithAnnotation
            .filter(SupportedElementFilter)
            .map(ElementToTypeAndFilePairFunction())
            .doOnNext { typeElementFileEntry ->
                MessagerUtils.reportInfo("Processing file: ${typeElementFileEntry.second}")
            }
            .map { it.first to it.second }
            .map(GenerateFileStreamTypeSpecFunction)
            .toList()

        if (fileStreamTypeSpecs.isNotEmpty()) {
            fileStreamTypeSpecs.run(GenerateMezzanineTypeSpecFunction)
                .run(TypeSpecToJavaFileFunction)
                .writeTo(codeGenerator, true)
        }

        return emptyList()
    }
}
